mitodl/ocw-hugo-themes
# fix hot reload in dev

#### Pre-Flight checklist


#### What are the relevant tickets?
Closes https://github.com/mitodl/ocw-hugo-themes/issues/806

#### What's this PR do?
In https://github.com/mitodl/ocw-hugo-themes/pull/787, we upgraded the version of webpack used in our webpack build to version 5 along with a number of dependencies.  A side effect of that is that when running the new dev server, if you make a change to a JS or CSS file, webpack will generate a "hot update" file.  This has the side effect of making the reference in the `.Site.Data.webpack` (`webpack.json`) data template an array instead of a single string.  So, when Hugo goes to render this it throws an error about expecting a string but getting an array:

```
[start:course:hugo] Change detected, rebuilding site.
[start:course:hugo] 2022-08-01 10:04:58.708 -0400
[start:course:hugo] Data changed "/home/gumaerc/Code/ocw-hugo-themes/base-theme/data/webpack.json": WRITE
[start:course:hugo] ERROR 2022/08/01 10:04:58 Rebuild failed:
[start:course:hugo] 
[start:course:hugo] ERROR 2022/08/01 10:04:58 Failed to render pages: render of "home" failed: execute of template failed: template: home.html:68:18: executing "home.html" at <partial "webpack_url.html" $theme.js>: error calling partial: "/home/gumaerc/Code/ocw-hugo-themes/base-theme/layouts/partials/webpack_url.html:3:11": execute of template failed: template: partials/webpack_url.html:3:11: executing "partials/webpack_url.html" at <hasPrefix . "/">: error calling hasPrefix: unable to cast []interface {}{"/static/js/main.js", "/static/main.3b26c1d8902715b1ddb7.hot-update.js"} of type []interface {} to string
```

This PR adds two new partials, `include_js.html` and `include_css.html`, and sets all the places including these files from a `webpack.json` entry to pass the data to these.  The partials check to see if the data is an array or not, and if it is, only include the first item in the list.

#### How should this be manually tested?
 - Read the readme if you've never run a site before locally.
 - Run `npm run start:www`
 - Visit http://localhost:3000 and verify that you see the `ocw-www` home page and there are no errors in the console
 - In `www/assets/css/home.scss`, add the following CSS rule:
```
body {
  background-color: red !important;
}
```
- Save the file and watch your browser window to make sure it properly reloads.  You should see a red background behind the promo carousel and there should be no errors in the console.
- In `www/assets/www.tsx`, under `initSubNav()` add `console.log("www test")` and save the file.  Your browser window should reload and in the console you should see the log entry.
- Stop the dev server, revert the changes you've made and run `npm run start:course`
- Verify that you see a course home page at http://localhost:3000 and there are no errors in the console.
- At the top of `course/assets/css/course-home.scss`, add the following:
```
#main-content {
  background-color: yellow;
}
```
- Save the file and watch your browser window.  It should reload and you should see a yellow background.  There should be no errors in the console.
- In `course/assets/course.ts`, under `initVideoFullscreenToggle()` add `console.log("course test")` and save the file.  Your browser window should reload and in the console you should see the log entry.

#### Where should the reviewer start?
`include_js.html`, `include_css.html` and the files that call them
